src/iceberg/catalog/memory_catalog.h

```diff
@@ -0,0 +1,204 @@
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+#pragma once
+
+#include <mutex>
+#include <optional>
+#include <unordered_map>
+
+#include "iceberg/catalog.h"
+
+namespace iceberg {
+
+class InMemoryNamespace;
+
+class ICEBERG_EXPORT InMemoryCatalog : public Catalog {
+ public:
+  InMemoryCatalog(std::shared_ptr<FileIO> file_io, std::string warehouse_location);
+
+  void Initialize(
+      const std::string& name,
+      const std::unordered_map<std::string, std::string>& properties) override;
+
+  std::string_view name() const override;
+
+  Result<std::vector<TableIdentifier>> ListTables(const Namespace& ns) const override;
+
+  Result<std::unique_ptr<Table>> CreateTable(
+      const TableIdentifier& identifier, const Schema& schema, const PartitionSpec& spec,
+      const std::string& location,
+      const std::unordered_map<std::string, std::string>& properties) override;
+
+  Result<std::unique_ptr<Table>> UpdateTable(
+      const TableIdentifier& identifier,
+      const std::vector<std::unique_ptr<UpdateRequirement>>& requirements,
+      const std::vector<std::unique_ptr<MetadataUpdate>>& updates) override;
+
+  Result<std::shared_ptr<Transaction>> StageCreateTable(
+      const TableIdentifier& identifier, const Schema& schema, const PartitionSpec& spec,
+      const std::string& location,
+      const std::unordered_map<std::string, std::string>& properties) override;
+
+  Status TableExists(const TableIdentifier& identifier) const override;
+
+  Status DropTable(const TableIdentifier& identifier, bool purge) override;
+
+  Result<std::shared_ptr<Table>> LoadTable(
+      const TableIdentifier& identifier) const override;
+
+  Result<std::shared_ptr<Table>> RegisterTable(
+      const TableIdentifier& identifier,
+      const std::string& metadata_file_location) override;
+
+  std::unique_ptr<iceberg::TableBuilder> BuildTable(const TableIdentifier& identifier,
+                                                    const Schema& schema) const override;
+
+ private:
+  std::string catalog_name_;
+  std::unordered_map<std::string, std::string> properties_;
+  std::shared_ptr<FileIO> file_io_;
+  std::string warehouse_location_;
+  std::unique_ptr<InMemoryNamespace> root_namespace_;
+  mutable std::recursive_mutex mutex_;
+};
+
+/**
+ * \brief A hierarchical namespace that manages namespaces and table metadata in-memory.
+ *
+ * Each InMemoryNamespace represents a namespace level and can contain properties,
+ * tables, and child namespaces. This structure enables a tree-like representation
+ * of nested namespaces.
+ */
+class ICEBERG_EXPORT InMemoryNamespace {
```

Comment:
I was simply mirroring the Java impl when I added the `Catalog` without notice of `SupportsNamespace`: https://github.com/apache/iceberg/blob/main/api/src/main/java/org/apache/iceberg/catalog/SupportsNamespaces.java. Does it make sense to add all interfaces from `SupportsNamespace` to our `Catalog` API? `iceberg-python` did the same thing: https://github.com/apache/iceberg-python/blob/main/pyiceberg/catalog/__init__.py#L588 This aligns with the rest catalog api.

@Fokko Do you have an opinion?